numpy/array_api/_dtypes.py

```diff
@@ -1,21 +1,42 @@
 import numpy as np
 
-# Note: we use dtype objects instead of dtype classes. The spec does not
-# require any behavior on dtypes other than equality.
-int8 = np.dtype("int8")
-int16 = np.dtype("int16")
-int32 = np.dtype("int32")
-int64 = np.dtype("int64")
-uint8 = np.dtype("uint8")
-uint16 = np.dtype("uint16")
-uint32 = np.dtype("uint32")
-uint64 = np.dtype("uint64")
-float32 = np.dtype("float32")
-float64 = np.dtype("float64")
-complex64 = np.dtype("complex64")
-complex128 = np.dtype("complex128")
+# Note: we wrap the NumPy dtype objects in a bare class, so that none of the
+# additional methods and behaviors of NumPy dtype objects are exposed.
+
+class _DType:
+    def __init__(self, np_dtype):
+        np_dtype = np.dtype(np_dtype)
+        self._np_dtype = np_dtype
+
+    def __repr__(self):
+        return f"np.array_api.{self._np_dtype.name}"
+
+    def __eq__(self, other):
```

Comment:
I'm thinking it might be a good idea to make `==` explicitly raise an exception when compared against a `numpy` dtype object. Right now `numpy.float32 == numpy.array_api.float32` returns `False`, but I feel like this could mask bugs. Is there a straightforward way to detect if `other` is a NumPy dtype object?

Reply:
This is a very fly-by comment, but in python generally `==` is supposed never to fail.

Reply:
Yes, I'm not completely sure about it. I did set this to error when I was creating this PR and it definitely helped me to find all the places in the code that needed to be updated. 

My worry if is someone is incorrectly doing:

```py
if x.dtype == np.float32: # x is a numpy.array_api array
    ...
```

or

```py
if x.dtype == 'f4':
    ...
```

The code will just subtly start being wrong.

But then again, I don't know if that kind of error actually exists in the wild. It is already wrong for other libraries like PyTorch.

OTOH, like you say, `==` failing can be very annoying as it would make things like `in` not work properly. I think I'm leaning towards not making this change, but I'm interested to hear what others think.

Reply:
Notwithstanding that this should move eventually.  For *this* namespace it seems right to raise errors liberally, since I think by now it is accepted as a *testing* implementation:  If it works with this, it should work with anything.

Reply:
How about a warning? For testing purposes, those can be turned into errors.

Reply:
I don't have an opinion, TBH.  Besides testing, I never saw much purpose to the namespace as is and in that context I think liberally raising is OK too.

OTOH, if you formalize the "this is for testing" part of the namespace, then it does seem quite logical to have a custom `UndefinedBehavior` warning.  That would make it easier to ignore it in tests if there is a reason to do so (not sure there is, but maybe it doesn't matter).